crates/prune/src/pruner.rs

```diff
@@ -0,0 +1,168 @@
+//! Support for pruning.
+
+use crate::PrunerError;
+use futures_util::Stream;
+use reth_primitives::BlockNumber;
+use reth_provider::CanonStateNotification;
+use std::{
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+};
+use tracing::debug;
+
+/// The future that returns the owned pipeline and the result of the pipeline run. See
+/// [Pruner::run_as_fut].
+pub type PrunerFut<St> = Pin<Box<dyn Future<Output = PrunerWithResult<St>> + Send>>;
+
+/// The pipeline type itself with the result of [Pruner::run_as_fut]
+pub type PrunerWithResult<St> = (Pruner<St>, Result<(), PrunerError>);
+
+/// Pruning routine. Main pruning logic happens in [Pruner::run].
+pub struct Pruner<St> {
+    /// Stream of canonical state notifications. Pruning is triggered by new incoming
+    /// notifications.
+    canon_state_stream: St,
```

Comment:
maybe we can make this boxed?
the generics are a bit painful for this

Reply:
right, agree